### ISSUE TITLE
CustomFromMask pruning stores a copy of the user-provided mask

### ISSUE THREAD

**Issue by mickypaganini** [pytorch/pytorch#34257]
## 🐛 Bug

(Not a bug, strictly speaking, because it doesn't compromise correctness, but still an issue because it affects model size).
The pruning method called `CustomFromMask` in `torch/nn/utils/prune.py` is designed to prune parameters by applying a user-provided mask. An instance of this class is then attached to the pruned module as a `forward_pre_hook`. The issue is that each hook carries (and therefore attaches to the module) a copy of the user-provided mask (of the same size as the pruned parameter), thus significantly growing the size of the model. This becomes particularly problematic in global pruning, which prunes the entire network by relying on `CustomFromMask`.

## To Reproduce

1. Create a `nn.Module`
1. Prune it using `torch.nn.utils.prune.custom_from_mask`
1. Check that a `forward_pre_hook` has been added to the Module
1. Notice that the hook carries and attribute called `mask` that stores a (undesirable) copy of the mask applied to the parameter that impacts the size of the Module, when serialized.

```python
import torch
from torch import nn
import torch.nn.utils.prune as prune

m = nn.Conv2d(3, 512, 3)
custom_mask = torch.randint(0, 2, size=m.weight.shape)
prune.custom_from_mask(m, 'weight', mask=custom_mask)
print([v.mask for v in m._forward_pre_hooks.values()])
```
Specifically, when comparing the size of a module pruned with any other pruning technique versus a module pruned with `CustomFromMask`, the size difference is clearly visible.
```python
m = nn.Conv2d(3, 512, 3)
custom_mask = torch.randint(0, 2, size=m.weight.shape)
prune.custom_from_mask(m, 'weight', mask=custom_mask)
torch.save(m, 'custom.pth')

n = nn.Conv2d(3, 512, 3)
prune.l1_unstructured(n, 'weight', amount=0.5)
torch.save(n, 'l1.pth')
```
Then `ls -lht *.pth` yields:
```
173K Mar  3 15:39 l1.pth
281K Mar  3 15:39 custom.pth
```
## Expected behavior
The user-provided mask is used only once when `compute_mask` is called, so it could be manually deleted afterwards. If keeping the mask was desirable for some use cases, we can add a flag in the constructor to determine whether the mask should be stored or deleted after it's been used.

## Environment

PyTorch version: 1.5.0.dev20200303
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.5.0.dev20200303
[pip3] torchvision==0.6.0.dev20200303
[conda] Could not collect

**Comment by mruberry** [comment]
Thank you for reporting this issue, @mickypaganini. @albanD, maybe we need a module for hook-related issues, too?

**Comment by albanD** [comment]
I don't think this is really hook related but the way pruning works.
The attached PR does solve the issue. I'll see with @mickypaganini to merge it.

**Comment by kkoutini** [comment]
Hi, any updated on the fix? 
I use iterative pruning, and this bug results in keeping track of all the masks' history.  I run out of gpu memory after some iterations. 

my use case is simple: call `prune.global_unstructured` iteratively. where `custom_from_mask(module, name, param_mask)` is getting called for each module. 

This will create a hook of `CustomFromMask` in the first call with an attribute `mask`. in the following calls it will create a hookof type `PruningContainer` with attribute `_pruning_methods` containing a list `CustomFromMask` hooks, each has a reference to a mask, and there is the memory leak.

If somebody having the same problem,  I made a dirty fix, by copying the `global_unstructured` function and deleting the masks from the hooks, after calling ` custom_from_mask(module, name, param_mask)`:
```python
        custom_from_mask(module, name, param_mask)
        for k in list(module._forward_pre_hooks):
            hook = module._forward_pre_hooks[k]
            if isinstance(hook, PruningContainer):
                if isinstance(hook[-1], CustomFromMask):
                    hook[-1].mask = None
            elif isinstance(hook, CustomFromMask):
                hook.mask = None

```
